doc/source/custom_directives.py

```diff
@@ -0,0 +1,217 @@
+
+        try:
+            text = open(filename).read()
+            text_no_docstring = self.docstring_regex.sub('', text, count=1)
+
+            code_block = nodes.literal_block(text=text_no_docstring)
+            return [code_block]
+        except FileNotFoundError as e:
+            print(e)
+            return []
+
+
+class GalleryItemDirective(Directive):
+    """
+    Create a sphinx gallery thumbnail for insertion anywhere in docs.
+
+    Optionally, you can specify the custom figure and intro/tooltip for the
+    thumbnail.
+
+    Example usage:
+
+    .. galleryitem:: intermediate/char_rnn_generation_tutorial.py
+        :figure: _static/img/char_rnn_generation.png
+        :intro: Put your custom intro here.
+
+    If figure is specified, a thumbnail will be made out of it and stored in
+    _static/thumbs. Therefore, consider _static/thumbs as a 'built' directory.
+    """
+
+    required_arguments = 1
+    optional_arguments = 0
+    final_argument_whitespace = True
+    option_spec = {
+        "figure": directives.unchanged,
+        "intro": directives.unchanged
+    }
+    has_content = False
+    add_index = False
+
+    def run(self):
+        args = self.arguments
+        fname = args[-1]
+
+        env = self.state.document.settings.env
+        fname, abs_fname = env.relfn2path(fname)
+        basename = os.path.basename(fname)
+        dirname = os.path.dirname(fname)
+
+        try:
+            if "intro" in self.options:
+                intro = self.options["intro"][:195] + "..."
+            else:
+                _, blocks = sphinx_gallery.gen_rst.split_code_and_text_blocks(
+                    abs_fname)
+                intro, _ = sphinx_gallery.gen_rst.extract_intro_and_title(
+                    abs_fname, blocks[0][1])
+
+            thumbnail_rst = sphinx_gallery.backreferences._thumbnail_div(
+                dirname, basename, intro)
+
+            if "figure" in self.options:
+                rel_figname, figname = env.relfn2path(self.options["figure"])
+                save_figname = os.path.join("_static/thumbs/",
+                                            os.path.basename(figname))
+
+                try:
+                    os.makedirs("_static/thumbs")
+                except OSError:
+                    pass
+
+                sphinx_gallery.gen_rst.scale_image(figname, save_figname, 400,
+                                                   280)
+                # replace figure in rst with simple regex
+                thumbnail_rst = re.sub(r"..\sfigure::\s.*\.png",
+                                       ".. figure:: /{}".format(save_figname),
+                                       thumbnail_rst)
+
+            thumbnail = StringList(thumbnail_rst.split("\n"))
+            thumb = nodes.paragraph()
+            self.state.nested_parse(thumbnail, self.content_offset, thumb)
+
+            return [thumb]
+        except FileNotFoundError as e:
```

Comment:
why do we have to wrap all of the above logic in this? shouldn't it just be where we're opening a file?